### PR TITLE
Add x-checker-data for kolibri-explore-plugin module

### PR DESCRIPTION
**Pull request by starnight**
Fixes: https://github.com/endlessm/endless-key-flatpak/issues/45

### CODE REVIEW

**Comment by dylanmccall** [comment]
Thanks for taking care of this, @StarNight! I pushed a small change which uses the JSON checker against https://api.github.com/repos/endlessm/kolibri-explore-plugin/releases/latest to get the newest download files. That checker is documented over here: https://github.com/flathub/flatpak-external-data-checker#json-checker.

That works nicely for me, now, but I'm wondering what you all think about getting this to fetch the whl from github as well? We do have it there, and I feel like fetching the whl from pypi and everything else from gihub introduces the possibility of things getting out of sync.

**Comment by starnight** [comment]
> It works nicely for me with that change, but I'm wondering what you all think about getting this to fetch the whl from github as well? We do have it there, and I feel like fetching the whl from pypi and everything else from github introduces the possibility of things getting out of sync.

Good question!  However, `kolibri_explore_plugin-7.1.0-py3-none-any.whl` includes the version string in the release page.  Besides, the wheel, apps-bundle and loading-screen are released in the same [release job](https://github.com/endlessm/kolibri-explore-plugin/blob/910e8b3e108ba8f7a3ed9e4a12cda557615ed1d2/.github/workflows/release.yml#L58).  It might be okay.

**Comment by starnight** [comment]
Ok!  I found a way to deal the version string in `kolibri_explore_plugin-7.1.0-py3-none-any.whl`.  Updated the commit.

Besides, I also added `x-checker-data` into other modules, too.  But, skip **kolibri** module.  Because, Endless Key uses kolibri's pre-release right now.

**Comment by dbnicholson** [comment]
> It works nicely for me with that change, but I'm wondering what you all think about getting this to fetch the whl from github as well? We do have it there, and I feel like fetching the whl from pypi and everything else from github introduces the possibility of things getting out of sync.

I've definitely thought about it a lot, but I think it's probably fine. Certainly they're the same file so long as no one is doing a manual upload to PyPI.

**Comment by dbnicholson** [comment]
One thing I'd really like to see here is version constraints. Should they happen, we would not want to automatically upgrade to kolibri-explore-plugin 8.y or kolibri 0.17.z.

**Review by dbnicholson — CHANGES_REQUESTED** [review]
I'm certainly in agreement with using this, but I'm not as excited about blindly updating major versions of transitive dependencies.

File: build-aux/flatpak/modules/iproute2.json [hunk 1]
```diff
@@ -8,9 +8,13 @@
...
-            "type" : "archive",
-            "url" : "https://mirrors.edge.kernel.org/pub/linux/utils/net/iproute2/iproute2-5.7.0.tar.xz",
-            "sha256" : "725dc7ba94aae54c6f8d4223ca055d9fb4fe89d6994b1c03bfb4411c4dd10f21"
+            "type" : "git",
+            "url" : "https://git.kernel.org/pub/scm/network/iproute2/iproute2.git",
+            "tag" : "v5.7.0",
+            "x-checker-data" : {
+                "type" : "git",
+                "tag-pattern" : "^v([\\d.]+)$"
+            }
```
[the remark this answers is not in the document]
Comment (starnight) [reply build-aux/flatpak/modules/iproute2.json:17]: I tested with iproute2 6.5.0.  Endless Key works as well.
Kolibri depends on [ifcfg](https://github.com/learningequality/kolibri/blob/v0.16.0-beta5/requirements/base.txt#L29) and the ifcfg package uses `ip` command from iproute2.

File: build-aux/flatpak/modules/python3-kolibri-app-desktop-xdg-plugin.json [hunk 2]
```diff
@@ -8,12 +8,21 @@
         {
             "type": "file",
             "url": "https://files.pythonhosted.org/packages/90/d4/a7c9b6c5d176654aa3dbccbfd0be4fd3a263355dc24122a5f1937bdc2689/Pillow-8.3.2.tar.gz",
-            "sha256": "dde3f3ed8d00c72631bc19cbfff8ad3b6215062a5eed402381ad365f82f0c18c"
+            "sha256": "dde3f3ed8d00c72631bc19cbfff8ad3b6215062a5eed402381ad365f82f0c18c",
+            "x-checker-data": {
+                "type": "pypi",
+                "name": "Pillow"
```
[the remark this answers is not in the document]
Comment (starnight) [reply build-aux/flatpak/modules/python3-kolibri-app-desktop-xdg-plugin.json:14]: Notice the restriction `Pillow>=8.2.0,<9.0.0` in [kolibri-app-desktop-xdg-plugin](https://github.com/endlessm/kolibri-app-desktop-xdg-plugin/blob/v1.1.4/setup.py#L28).

File: build-aux/flatpak/modules/python3-virtualenv-api.json [hunk 3]
```diff
@@ -8,7 +8,12 @@
         {
             "type": "file",
             "url": "https://files.pythonhosted.org/packages/ee/ff/48bde5c0f013094d729fe4b0316ba2a24774b3ff1c52d924a8a4cb04078a/six-1.15.0-py2.py3-none-any.whl",
-            "sha256": "8b74bedcbbbaca38ff6d7491d76f2b06b3592611af620f8426e82dddb04a5ced"
+            "sha256": "8b74bedcbbbaca38ff6d7491d76f2b06b3592611af620f8426e82dddb04a5ced",
+            "x-checker-data": {
+                "type": "pypi",
+                "name": "six",
```
[the remark this answers is not in the document]
Comment (starnight) [reply build-aux/flatpak/modules/python3-virtualenv-api.json:14]: I have tested without `modules/python3-virtualenv-api.json`.  Endless Key builds and works as usual.  I am going to drop `modules/python3-virtualenv-api.json`.  Please blame me if it is needed.

**Comment by starnight** [comment]
Also, add `x-checker-data` into kolibri module.

File: build-aux/flatpak/modules/python3-kolibri-app-desktop-xdg-plugin.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (starnight) [reply build-aux/flatpak/modules/python3-kolibri-app-desktop-xdg-plugin.json]: Updated the commit.

File: build-aux/flatpak/modules/python3-kolibri-explore-plugin.json [hunk 4]
```diff
@@ -10,21 +10,42 @@
...
             "url": "https://github.com/endlessm/kolibri-explore-plugin/releases/download/v7.1.0/apps-bundle.zip",
             "sha256": "ca771b748fb10d977ef2c0f1b651687eaf4696f789f7bd6db524185fa50fe814",
-            "dest": "apps-bundle"
+            "dest": "apps-bundle",
+            "x-checker-data": {
+                "type": "json",
+                "url": "https://api.github.com/repos/endlessm/kolibri-explore-plugin/releases/latest",
+                "version-query": ".tag_name",
+                "url-query": ".assets[] | select(.name==\"apps-bundle.zip\") | .browser_download_url",
+                "versions": {"<": "8.0.0"}
```
[the remark this answers is not in the document]
Comment (dbnicholson) [reply build-aux/flatpak/modules/python3-kolibri-explore-plugin.json:33]: As it turns out, the JSON checker doesn't support `versions`, so we'll just have to do this with `jq` filters.

**Review by dbnicholson — APPROVED** [review]
This is good enough to start even though some of the version constraints won't work. I rebased to address the conflict and added the `v` stripping for the apps-bundle and loading-screen versions.

File: build-aux/flatpak/modules/python3-kolibri.json [hunk 5]
```diff
@@ -9,7 +9,14 @@
         {
             "type": "file",
             "url": "https://github.com/learningequality/kolibri/releases/download/v0.16.0-beta5/kolibri-0.16.0b5-py2.py3-none-any.whl",
-            "sha256": "3925b7a18b6684547fd50e1df7326ace318d9d81d1fa02317f1870e117eb515c"
+            "sha256": "3925b7a18b6684547fd50e1df7326ace318d9d81d1fa02317f1870e117eb515c",
+            "x-checker-data": {
+                "type": "json",
+                "url": "https://api.github.com/repos/learningequality/kolibri/releases",
+                "version-query": "first | .tag_name | sub(\"^[vV]\"; \"\") | sub(\"-beta\"; \"b\")",
+                "url-query": "first | .assets[] | select(.name==\"kolibri-\" + $version + \"-py2.py3-none-any.whl\") | .browser_download_url",
```
[the remark this answers is not in the document]
Comment (starnight) [reply build-aux/flatpak/modules/python3-kolibri.json:17]: The `/latest` only shows the latest normal release.  But, we are using Kolibri's **pre-release**, which will not be shown by `/latest`.  So, we can only pick the first release by ourselves.

**Pull request merged**